File: 30-04-2024/Add two numbers represented by linked lists.js
//{ Driver Code Starts
//Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

class Node{
    constructor(data){
        this.data = data;
        this.next = null;
    }
}

function printlist(head){
    let current = head; 
    let s='';
    while (current !== null)
    {
        s+=current.data+" ";
        current = current.next;
    }
    console.log(s);
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let n = parseInt(readLine());
        let input_ar1 = readLine().split(' ').map(x=>parseInt(x));
        let num1 = new Node(input_ar1[0]);
        let tail1 = num1;
        for(let i=1;i<n;i++){
            tail1.next = new Node(input_ar1[i]);
            tail1 = tail1.next;
        }
        
        let m = parseInt(readLine());
        let input_ar2 = readLine().split(' ').map(x=>parseInt(x));
        let num2 = new Node(input_ar2[0]);
        let tail2 = num2;
        for(let i=1;i<m;i++){
            tail2.next = new Node(input_ar2[i]);
            tail2 = tail2.next;
        }
        
        let obj = new Solution();
        let res = obj.addTwoLists(num1, num2);
        printlist(res);
    }
}
// } Driver Code Ends




//User function Template for javascript

/**
 * @param {Node} first
 * @param {Node} second
 * @returns {Node}
*/

/*
class Node{
    constructor(data){
        this.data = data;
        this.next = null;
    }
}
*/
//Your Code Start
class Solution {
    //Function to add two numbers represented by linked list.
    addTwoLists(num1, num2)
    {
        const reverse = (head) => {
            let prev = null;
            let curr = head;
            while (curr) {
                let nextNode = curr.next;
                curr.next = prev;
                prev = curr;
                curr = nextNode;
            }
            return prev;
        };

        num1 = reverse(num1);
        num2 = reverse(num2);
        let result = null;
        let temp = null;
        let carry = 0;
        while (num1 || num2 || carry) {
            let sumVal = carry;
            if (num1) {
                sumVal += num1.data;
                num1 = num1.next;
            }
            if (num2) {
                sumVal += num2.data;
                num2 = num2.next;
            }
            carry = Math.floor(sumVal / 10);
            sumVal = sumVal % 10;
            let newNode = new Node(sumVal);
            if (!result) {
                result = newNode;
            } else {
                temp.next = newNode;
            }
            temp = newNode;
        }
        if (carry > 0) {
            temp.next = new Node(carry);
        }
        result = reverse(result);
        while (result && result.data === 0 && result.next) {
            result = result.next;
        }
        return result;
    }
}
//Your Code End

File: 30-04-2024/Add two numbers represented by linked lists.py
# Node for linked list:
#Your Code Start
class Solution:
    # Function to reverse a linked list
    @staticmethod
    def reverse(head):
        prev = None
        curr = head
        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        return prev

    # Function to add two numbers represented by linked list
    @staticmethod
    def addTwoLists(num1, num2):
        num1 = Solution.reverse(num1)
        num2 = Solution.reverse(num2)
        result = None
        temp = None
        carry = 0
        while num1 or num2 or carry:
            sum_val = carry
            if num1:
                sum_val += num1.data
                num1 = num1.next
            if num2:
                sum_val += num2.data
                num2 = num2.next
            carry = sum_val // 10
            sum_val = sum_val % 10
            new_node = Node(sum_val)
            if not result:
                result = new_node
            else:
                temp.next = new_node
            temp = new_node
        if carry > 0:
            temp.next = Node(carry)
        result = Solution.reverse(result)
        while result and result.data == 0 and result.next:
            result = result.next
        return result

#Your Code End



#{ 
 # Driver Code Starts
#Initial Template for Python 3

# Node Class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Linked List Class
class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    # creates a new node with given value and appends it at the end of the linked list
    def insert(self, val):
        if self.head is None:
            self.head = Node(val)
            self.tail = self.head
        else:
            self.tail.next = Node(val)
            self.tail = self.tail.next

# prints the elements of linked list starting with head
def printList(n):
    while n:
        print(n.data,end=' ')
        n = n.next
    print()

if __name__ == '__main__':
    for _ in range(int(input())):
        
        n = int(input())
        arr1 = ( int(x) for x in input().split() )
        num1 = LinkedList()
        for i in arr1:
            num1.insert(i)
        
        m = int(input())
        arr2 = ( int(x) for x in input().split() )
        num2 = LinkedList()
        for i in arr2:
            num2.insert(i)
        
        res = Solution().addTwoLists(num1.head, num2.head)
        printList(res)
# } Driver Code Ends
